contracts/contracts/Multisig.cairo

```diff
@@ -470,24 +432,119 @@ func execute_transaction{
     return (response_len=response.retdata_size, response=response.retdata)
 end
 
+# Sets number of required confirmations. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_confirmations_required.
+@external
+func set_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    confirmations_required : felt
+):
+    require_multisig()
+
+    const lower_bound = 1
+
+    let (owners_len) = _owners_len.read()
+    with_attr error_message("invalid number of required confirmations"):
+        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
+    end
+
+    _set_confirmations_required_impl(confirmations_required)
+
+    return ()
+end
+
+# Sets the owners field on the multisig. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_owners.
+# Number of required confirmations is decreased in case its larger than
+# number of new owners.
+@external
+func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    require_multisig()
+    _set_owners_impl(owners_len, owners)
+
+    return ()
+end
+
+# Set new owners and number of required confirmations.
+# Can be called only via recursively from
+# execute_transaction -> set_owners_and_confirmations_required
+@external
+func set_owners_and_confirmations_required{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(owners_len : felt, owners : felt*, confirmations_required : felt):
+    set_owners(owners_len, owners)
```

Comment:
Sure this can be done via external call there is `call_contract` that can make recursive calls. But out curiosity: what is the reasoning behind making external call here? To avoid argument allocation in memory?

Reply:
So my comment is suggesting the current workflow might produce external instead of internal (preferred, cheaper) calls. That's why I suggested functions should be refactored to make sure these calls are internal when calling `set_owners_and_confirmations_required` fn.

Reply:
We can resolve this comment and focus on the new code structure discussion that will also fix this issue: https://github.com/eqlabs/starknet-multisig/pull/22#discussion_r915872245